src/lib.rs

```diff
@@ -2228,6 +2227,27 @@ macro_rules! impl_boilerplate_for {
                 Ok(Self::DEVICE_TYPE)
             }
         }
+
+        impl MpsseCmdExecutor for $DEVICE {
+            type Error = TimeoutError;
+
+            fn mpsse_init(&mut self, settings: &MpsseSettings) -> Result<(), Self::Error> {
+                self.initialize_mpsse(settings)
+            }
+
+            fn mpsse_send(&mut self, data: &[u8]) -> Result<(), Self::Error> {
+                self.write_all(data)
+            }
+
+            fn mpsse_xfer(&mut self, txdata: &[u8], rxdata: &mut [u8]) -> Result<(), Self::Error> {
+                self.write_all(txdata)?;
+                self.read_all(rxdata)
+            }
+
+            fn mpsse_close(&mut self) {
+                self.close().ok();
+            }
+        }
```

Comment:
1. Concerning read_all/write_all
Could you please clarify what you mean ?
Am I correct assuming that you suggest to define MpsseCmdExecutor for std::io::{Read,Write} implementors e.g. along the following lines:
```
pub trait MpsseCmdExecutor: std::io::Read + std::io::Write {
    /// Error type
    type Error;

    /// Configure FTDI MPSSE mode
    fn mpsse_init(&mut self, settings: &MpsseSettings) -> Result<(), Self::Error>;

    /// Execute MPSSE command sequence
    fn mpsse_send(&mut self, data: &[u8]) -> Result<(), Self::Error> {
        self.write_all(data)
    }

    /// Execute MPSSE command and read response
    fn mpsse_xfer(&mut self, txdata: &[u8], rxdata: &mut [u8]) -> Result<(), Self::Error> {
        self.write_all(txdata)?;
        self.read_all(rxdata)
    }

    /// Close FTDI MPSSE device
    fn mpsse_close(&mut self);
}
```

2. On closing mpsse
This is a very good question. In ```ftdi-embedded-hal``` I create HAL handler from device handler that should implement MpsseCmdExecutor. I pass that device handler to HAL by value rather than by reference in order to avoid too many lifetime annotations internally. So that device is consumed by HAL. However after I am done with that HAL instance and  e.g. it leaves the scope, I still can not create another ```libftd2xx-rs``` device instance in the same process for the same device unless I explicitly close previous device instance. This problem does not exist for ```ftdi-rs``` device handlers. Probably  ```libftd2xx``` C library tracks per-process state internally. I tried to implement Drop in ```libftd2xx-rs```, but that approach does not work well with passing device handler into HAL by value because device handler is re-created at some point loosing valid FT_HANDLE value. So for now I use this ```close``` method. Any suggestions are welcome.

3. FtdiMpsse methods
See my comment for 1. Though I don't quite understand how to define default method for mpsse_init and mpsse_close.

Reply:
I figured out that my explanation about ```mpsse_close``` is a bit fuzzy. So here are some details and clarifications. I will start from a simple test program:

```rust
use libftd2xx::{Ft232h, FtdiCommon};
use std::time::Duration;

fn main() {
    let ft = Ft232h::with_description("Single RS232-HS").unwrap();
    test(ft, "T1");

    let ft = Ft232h::with_description("Single RS232-HS").unwrap();
    test(ft, "T2");
}   

fn test(mut d: Ft232h, t: &str) {
    println!("{}: TP#1", t);
    d.set_latency_timer(Duration::from_millis(10)).unwrap();
    //d.close().unwrap();
    println!("{}: TP#2", t);
}
```

The second ```with_description``` call fails with ```DEVICE_NOT_OPENED``` if device is not properly closed. Removing comment from the ```close``` call makes things work.

Another option would be to implement ```Drop``` trait. The question is which type should implement it. If ```Drop``` trait is implemented for ```Ftdi```, then it will be called in ```try_from``` when ```Ftdi``` device is promoted to ```Ft232h```. As a result, the above code will fail with ```INVALID_HANDLE``` right on the first ```with_description``` call. Implementing ```Drop``` trait for ```Ftx232h``` types fixes this simple code. However it does not look consistent in general. So I added ```mpsse_close``` method as a temporary fix until proper solution is found.

Maybe ```Ftx232h``` types should embed ```Ftdi``` handler rather than drop it after copying internal handler ?

Reply:
New RFC commit has been added to MR: Ftdi embedding and Drop implementation. Now the above example works fine w/o additional close. So mpsse_close is not needed anymore.

Reply:
> Concerning read_all/write_all Could you please clarify what you mean ?

Sorry, I was not very clear.
The method `mpsse_xfer` could be derived with a provided `read_all` and `write_all` function; should the methods that are required simply be a `read_all` and `write_all` with a default implementation for `mpsse_xfer`?  A lot of IO traits I see abstract only the `read`/`write` portions.  I am ok with it either way.

In addition, should some of the methods from `FtdiMpsse` be moved into this trait?  For example `clock_data` only needs `mpsse_send` and `mpsse_xfer` so it could be implemented with a default method in this trait.
https://github.com/newAM/libftd2xx-rs/blob/e3e7eaeb9a603d92a257bf61314a4137318b1c8a/src/mpsse.rs#L421-L429

> New RFC commit has been added to MR:
> 
>     * embedding Ftdi into Ftx232h
> 
>     * implement Drop for Ftdi
> 
> 
> Now the above example works fine w/o additional close operation. So `mpsse_close` is not needed as well.
> If this or similar change will be done, then `mpsse_close` can be removed from MpsseCmdExecutor trait.

That works for me!

Reply:
Ok, thanks for clarification. That looks reasonable. So I updated MpsseCmdExecutor trait definition in ```ftdi-mpsse```. Now there are 3 basic methods: ```init```/```send```/```recv```. Other methods, if needed, can be derived from ```send```/```recv```. For instance, default implementation of ```xfer``` method has been added. Appropriate change has been also added to this PR.

Reply:
As you mentioned, some other basic methods can be moved to ```MpsseCmdExecutor``` trait default implementations. It looks like there are two consistent options: either almost all those methods (clock, loopback, gpio, clock_data) should be added to the trait or none of them. Meanwhile often those commands are combined into sequences. For insance  there are common patterns like appending ```send_immediate```. So for now I would suggest to keep trait as small as possible, leaving command construction and execution to the end user.

Reply:
BTW, if you are ok with the general direction, then could you please create ```ftdi-mpsse``` empty project repository in your github workspace. I will create PR from the current development [repository](https://github.com/geomatsi/ftdi-mpsse).

Reply:
This is adding up to a lot of repos, so I created a github organization: https://github.com/ftdi-rs
I sent you an invite.

I think it makes sense to move `libftd2xx`, `libftd2xx-ffi`, `ftdi-mpsse`, and hal crates under this org.

Maybe we can get `libftdi1-sys` and `libftdi` under this org as well (if the maintainers agree)

Reply:
I saw the invite accept, I made you an owner so that you can create repos in the organization

Reply:
Thanks. I joined to ftdi-rs. So I will create ```ftdi-mpsse```. Later on, as soon as we are done with the reviews for this crate and ```ftdi-mpsse```, I will move ```ftdi-embedded-hal``` there as well.

Reply:
> So for now I would suggest to keep trait as small as possible, leaving command construction and execution to the end user.

Sorry just read all the other comments, they were hidden under a drop down.

This sounds good to me, the traits can always be expanded with default members as a non-breaking change later.